### DISCUSSION TITLE
PyRx based extension

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#93] [Ideas]
I'm thinking about this scenario.
Based on `PyRx`, I created an extension for Autocad for which I would like to create a simple installer.
In the global directory (`%localappdata%\Programs\PyRx`) I would have to install everything related to `PyRx` (skip this step if it is already installed).
I would like my extension to be able to be added as automatically loaded when CAD starts.
I wouldn't want to modify the global `pyrx_onload.py` for this.
I also cannot (under current assumptions) add my own `pyrx_onload.py` because `PyRx.arx` may already be loaded and then my  `pyrx_onload.py` will not be loaded.

I see two solutions:
1. Load the `pyrx_onload.py` file from `PyRxLoader.arx`, then just rename it to `MyExtName.arx` - my `pyrx_onload.py` file will be loaded, and `PyRx.arx` will be skipped because it is already loaded,
2. Establish some procedure based on which other extensions are loaded when loading `PyRx.arx`, e.g. registry entries pointing to `pyrx_onload.py` files of specific extensions.

**Comment by CEXT-Dan** [comment]
Not sure I have any advice, there’s lots of ways you can load modules through one of the AutoLISP mechanisms
https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/Automatically-load-AutoLISP-routines.html

I think you can use a .bundle too. you just have to make sure python is loaded early. 

pyrx_onload.py was intended for this, the installer no longer replaces or installs the file

**Reply by gswifort** [reply]
You're right, that should be enough.

**Reply by gswifort** [reply]
I tried to use [solution 1](https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/Automatically-load-AutoLISP-routines.html#:~:text=Solution%3A-,Add%20to%20the%20Startup,-Run%20the%20APPLOAD) provided by Autodesk (I use ZWCAD).
The problem occurs when `PyRx` is not loaded, then I load `RxLoader`, but after loading it, `PyRx` loads with a delay, so I cannot immediately invoke the `PYLOAD` command. Splitting it into two files didn't help either. I'm bad at AutoLisp, I have no other ideas.

Below is my broken code:

```lsp
; file 1
(defun c:LoadPyRx ()
  (progn
    (if (not (arxloaded "RxLoaderZ25.0.zrx"))
      (progn
        (setq PyRxLoaderPath (strcat (getenv "localappdata") "\\Programs\\PyRx\\Bin\\RxLoaderZ25.0.zrx"))
        (if (findfile PyRxLoaderPath)
          (arxload PyRxLoaderPath)
          (progn
            (princ "\nFile not found: PyRxLoader.arx")
            (exit)
          )
        )
      )
    )
)

(defun arxloaded (arxname)
  (if (member arxname (arx))
    T
    nil
  )
)

(c:LoadPyRx)

```

```lsp
; file 2
(defun c:LoadPyRx2 ()
  (progn
    (if (not (arxloaded "pyrxz25.0.zrx"))
      (progn
        (princ "\nNot loaded: PyRxZ25.0.zrx")
        (exit)
      )
      (progn
        (command "FILEDIA" "0")
        (command "PYLOAD" "C:\\Users\\gswi\\AppData\\Local\\Programs\\PyRx\\Bin\\pyrx_onload.py")
        (command "FILEDIA" "1")
      )
    )
  )
  (princ)
)

(c:LoadPyRx2)
```

**Reply by CEXT-Dan** [reply]
Oh yeah, originally Py_InitializeEx and wxPyGetAPIPtr crashed when called in On_kInitAppMsg, so I had to move it to OnIdleWinMsg, so there is a delay. I think the reason is either Python or wxPython won’t load in the UI thread.  

Since then, the loader module was added, so I can try moving the Py_InitializeEx back to On_kInitAppMsg, this will mean they the loader must always be used. I had considered that that someday the loader may not be necessary, if the proper environment was already setup 

There’s also a delay in pyrx_onload, it runs in the next on idle sequence after the first document is loaded. Also acedDefun fails if it’s run in the application context, ZwCAD is missing beginExecuteInCommandContext, so I have to add another hack lol.

The Loading sequence is very complex and the behavior is different between all the platforms.

Unfortuatly, if On_kInitAppMsg doesn't work, pyrx_onload may be your only option

**Reply by CEXT-Dan** [reply]
I’ve removed the delay, everything seems to work fine, all tests pass.
So maybe this will fix loading with lisp, I’ll publish a build

**Reply by gswifort** [reply]
It works! 
Although the order of the logs is not preserved (but that's no problem).

```raw
d:\Users\GSWI\Download\autoload.lsp load successfully!
Command: 
PyRx version <1.3.029.20240717> loaded:
Python Interpreter Loaded successfully!
FILEDIA

Enter FILEDIA's new value  <1>: 0
Command: PYLOAD
Select Python File: C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\pyrx_onload.py
Success module PYRX_ONLOAD is loaded: onload: C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\pyrx_onload.py
PyRxLoader::getPythonVenvPath 
PyRxLoader::thisModulePath C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\
PyRxLoader::getInstallPath C:\Users\gswi\AppData\Local\Programs\PyRx
PyRxLoader::getIniPath C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\PyRx.INI
PyRxLoader::validateINIPythonInstallPath c:\users\gswi\appdata\local\programs\python\python312
PyRxLoader::validateINIwxPythonPath c:\users\gswi\appdata\local\programs\python\python312\Lib\site-packages\wx
Loading PyRx from ini condition
PyRxLoader::PyRxLoader_loader Loading, C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\PyRxZ25.0.Zrx
Finished PyRxLoader_loader
Command: FILEDIA

Enter FILEDIA's new value  <0>: 1
```

**Reply by CEXT-Dan** [reply]
Yay!
